Separate monitor logic from transaction data

luckysori opened the pull request:
Fixes https://github.com/itchysats/itchysats/issues/3028.

When we worked on https://github.com/itchysats/itchysats/commit/b4225d5d35e27caef749d5c8ebf4e0f28c55d649 we inadvertently introduced a bug[1]. The gist of it is that, after a restart following a commit transaction being confirmed, we were not monitoring for the CET and refund timelocks anymore. This means that such a CFD would be stuck.

To fix it we have separated the transaction data that becomes available as events are processed from the logic of whether a
transaction (or a timelock) should be monitored or not. The data should always be remembered, and only the flags that control the monitoring logic should change as new events are processed.

[1]: https://github.com/itchysats/itchysats/issues/3028.

luckysori committed:
fix(monitor): Handle early states and RolloverRejected differently

When re-hydrating the `monitor::Cfd` we do not need to do anything to
the aggregate when handling `ContractSetupStarted`,
`ContractSetupFailed`, `OfferRejected`. These states can only occur
right after the creation of the CFD, so there is nothing that needs to
be enabled or disabled in terms of monitoring if contract setup didn't
succeed.

For `RolloverRejected`, we should definitely not disable things like
monitoring for the commit transaction finality. That could see us
failing to rollover and then not checking if the other party has
published the current commit transaction!

luckysori committed:
fix(monitor): Separate monitor logic from transaction data

When we worked on b4225d5d35e27caef749d5c8ebf4e0f28c55d649 we
inadvertently introduced a bug[1]. The gist of it is that, after a
restart following a commit transaction being confirmed, we were not
monitoring for the CET and refund timelocks anymore. This means that
such a CFD would be stuck.

To fix it we have separated the transaction data that becomes
available as events are processed from the logic of whether a
transaction (or a timelock) should be monitored or not. The data
should always be remembered, and only the flags that control the
monitoring logic should change as new events are processed.

[1]: https://github.com/itchysats/itchysats/issues/3028.

bonomat reviewed (approved):
Tested locally ☑️

da-kami commented on daemon/src/monitor.rs:
```diff
@@ -117,19 +117,29 @@ pub struct Actor {
-    commit_tx: Option<Transaction>,
+    broadcast_lock: Option<Transaction>,
+    broadcast_cet: Option<Transaction>,
+    broadcast_commit: Option<Transaction>,
```
❓ Why is there no `broadcast_refund`?

luckysori replied:
There was never a `refund: Option<Transaction>` as you can see on the left side of this diff.

I think we want to avoid situations where refund is published automatically after a restart. 

---

Aside from this PR:

To be honest, I'm not sure refund should be published automatically during runtime either. It's rare enough and disrupting enough (doesn't really enforce the bet) that I would argue it should always be triggered manually by either party if the oracle never does its thing. You could theoretically even let the parties settle collaboratively instead.

da-kami commented on daemon/src/monitor.rs:
```diff
@@ -117,19 +117,29 @@ pub struct Actor {
-    commit_tx: Option<Transaction>,
+    broadcast_lock: Option<Transaction>,
+    broadcast_cet: Option<Transaction>,
+    broadcast_commit: Option<Transaction>,
```
🤔 Unless I missed something `broadcast_commit` is never set to `Some`. Is that on purpose or is there something missing here?

luckysori replied:
Good catch! But looking at the diff we never actually set it to `Some` before either! Which means if we ever found ourselves in a situation where a restart occurred after a `ManualCommit` was emitted but the commit transaction publication triggered by the `process_manager` _failed_, we would have been stuck.

I'll add another commit to fix this.

luckysori replied:
Done with https://github.com/itchysats/itchysats/pull/3029/commits/63f0a89c801326e5fe3f045efb32d17e3a3c6d95.

da-kami commented on daemon/src/monitor.rs:
```diff
@@ -207,86 +222,70 @@ impl Cfd {
+                cet: cet_txid_and_script(cet),
+                monitor_cet_finality: true,
                 monitor_cet_timelock: false,
                 ..self
```
🤔 Given that we apply the events in order and some events set `broadcast_...` transactions, shouldn't *all* other events (that don't set `broadcast_...` to `Some`) set the value to `None`? 
Otherwise we might pass on `Some` values unintended.

luckysori replied:
I don't think so.

As an example, if we have a `ManualCommit` event which sets the `broadcast_commit` transaction to `Some` and we get an `OracleAttestedPriorCetTimelock`, we must still broadcast the commit transaction.

I think particularly with the rebroadcasting mechanism we don't need to be that careful anyway. If we fail to rebroadcast because it's already there or because something else ended up happening nothing bad can happen.

luckysori committed:
fix(monitor): Publish commit transaction on restart

We were never actually setting the `broadcast_commit` (the old
`commit_tx`) field to `Some`, so the commit transaction could never be
re-broadcast on a restart of the application.

The event `ManualCommit` is the only one that warrants the publication
of the commit transaction. Once we process a `CommitConfirmed` event
too, the transaction will not need to be re-broadcast, so we set
`broadcast_commit` to `None`.
